src/flint/types/fmpz_mod_poly.pyx

```diff
@@ -1536,6 +1536,213 @@ cdef class fmpz_mod_poly(flint_poly):
+            6*x^2 + 4*x + 2
+        """
+        # Only allow addition with other fmpz_mod_poly
+        if not typecheck(other, fmpz_mod_poly):
+            raise TypeError("other polynomial must be of type fmpz_mod_poly")
+
+        # Ensure the contexts match
+        other_c = <fmpz_mod_poly>other
+        if self.ctx != other_c.ctx:
+            raise ValueError("other polynomial's context does not match")
+
+        cdef fmpz_mod_poly res
+        res = self.ctx.new_ctype_poly()
+        fmpz_mod_poly_add_series(
+            res.val, self.val, other_c.val, n, res.ctx.mod.val
+        )
+        return res
+
+    def sub_trunc(self, other, slong n):
+        """
+        Truncate ``self`` and ``other`` to polynomials of length ``n`` and return their difference
+
+            >>> R = fmpz_mod_poly_ctx(163)
+            >>> f = R([2,3,5,7,11])
+            >>> h = R([1,2,4,8,16])
+            >>> f.sub_trunc(h, 2)
+            x + 1
+            >>> f.sub_trunc(h, 4)
+            162*x^3 + x^2 + x + 1
+        """
+        # Only allow subtraction with other fmpz_mod_poly
+        if not typecheck(other, fmpz_mod_poly):
+            raise TypeError("other polynomial must be of type fmpz_mod_poly")
+
+        # Ensure the contexts match
+        other_c = <fmpz_mod_poly>other
+        if self.ctx != other_c.ctx:
+            raise ValueError("other polynomial's context does not match")
+
+        cdef fmpz_mod_poly res
+        res = self.ctx.new_ctype_poly()
+        fmpz_mod_poly_sub_series(
+            res.val, self.val, other_c.val, n, res.ctx.mod.val
+        )
+        return res
+
+    def mul_low(self, other, slong n):
+        r"""
+        Returns the lowest ``n`` coefficients of the multiplication of ``self`` with ``other``
+
+        Equivalent to computing `f(x) \cdot g(x) \mod x^n`
+
+            >>> R = fmpz_mod_poly_ctx(163)
+            >>> f = R([2,3,5,7,11])
+            >>> g = R([1,2,4,8,16])
+            >>> f.mul_low(g, 5)
+            101*x^4 + 45*x^3 + 19*x^2 + 7*x + 2
+            >>> f.mul_low(g, 3)
+            19*x^2 + 7*x + 2
+            >>> f.mul_low(g, 1)
+            2
+        """
+        # Only allow multiplication with other fmpz_mod_poly
+        if not typecheck(other, fmpz_mod_poly):
+            raise TypeError("other polynomial must be of type fmpz_mod_poly")
+
+        # Ensure the contexts match
+        other_c = <fmpz_mod_poly>other
+        if self.ctx != other_c.ctx:
+            raise ValueError("other polynomial's context does not match")
+
+        cdef fmpz_mod_poly res
+        res = self.ctx.new_ctype_poly()
+        fmpz_mod_poly_mullow(
+            res.val, self.val, other_c.val, n, res.ctx.mod.val
+        )
+        return res
+
+    def mul_high(self, other, slong start):
+        r"""
+        Returns from coefficients from ``start`` onwards of the multiplication of ``self`` with ``other``
+
+        # TODO!!!
+        #
+        # Is this expected behaviour? Seems weird.
```

Comment:
The doc says:

> Computes the product of poly1 and poly2 and writes the coefficients from start onwards into the high coefficients of res, **the remaining coefficients being arbitrary.**

I think that the implementation just decided that it was faster to compute the full product rather than selectively compute the highest coefficients.

I guess we accept and document that or otherwise zero out the lower coefficients:
```python
In [23]: f.mul_high(g, 5)
Out[23]: 13*x^8 + 37*x^7 + 17*x^6 + 138*x^5 + 101*x^4 + 45*x^3 + 19*x^2 + 7*x + 2

In [24]: f.mul_high(g, 5).right_shift(5).left_shift(5)
Out[24]: 13*x^8 + 37*x^7 + 17*x^6 + 138*x^5
```

Reply:
yeah I think my feeling is if we have to implement `f.mul_high(g, 5).right_shift(5).left_shift(5)` on our side, we should just not have the method? Maybe the output is good for C flint, but for us the output seems not very interesting.

Reply:
Maybe it is not worth having. I can imagine wanting the product of the leading terms or the leading coefficients but I'm not sure I when I would particularly want what this function does. It does not look like `fmpz_mod_poly_mulhigh` is actually used anywhere in Flint itself.

It can also be implemented in a fairly straight-forward way by calling `right_shift` before multiplying.